File: mobile-flutter/test/models/app_state_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:careconnect/models/app_state.dart';

void main() {
  group('AppState — medications', () {
    late AppState state;
    setUp(() => state = AppState());

    test('addMedication appends and logs activity', () {
      state.addMedication(Medication(id: 'm1', name: 'Amlodipine', dose: '5mg', time: '8am', notes: ''));
      expect(state.medications, hasLength(1));
      expect(state.medications.first.name, 'Amlodipine');
      expect(state.activity, hasLength(1));
      expect(state.activity.first.type, ActivityType.taskCompleted);
    });

    test('toggleMedTaken flips taken and logs the correct activity type', () {
      state.addMedication(Medication(id: 'm1', name: 'Amlodipine', dose: '5mg', time: '8am', notes: ''));
      state.toggleMedTaken('m1');
      expect(state.medications.first.taken, isTrue);
      expect(state.activity.first.type, ActivityType.medicationTaken);

      state.toggleMedTaken('m1');
      expect(state.medications.first.taken, isFalse);
      expect(state.activity.first.type, ActivityType.medicationUnmarked);
    });

    test('deleteMedication removes the matching medication only', () {
      state.addMedication(Medication(id: 'm1', name: 'A', dose: '', time: '', notes: ''));
      state.addMedication(Medication(id: 'm2', name: 'B', dose: '', time: '', notes: ''));
      state.deleteMedication('m1');
      expect(state.medications, hasLength(1));
      expect(state.medications.first.id, 'm2');
    });

    test('toggleMedTaken on an unknown id is a no-op', () {
      state.addMedication(Medication(id: 'm1', name: 'A', dose: '', time: '', notes: ''));
      state.toggleMedTaken('does-not-exist');
      expect(state.medications.first.taken, isFalse);
    });
  });

  group('AppState — appointments', () {
    late AppState state;
    setUp(() => state = AppState());

    test('addAppointment inserts at the front', () {
      state.addAppointment(Appointment(id: 'a1', title: 'First', dateTime: '', location: '', notes: ''));
      state.addAppointment(Appointment(id: 'a2', title: 'Second', dateTime: '', location: '', notes: ''));
      expect(state.appointments.first.id, 'a2');
    });

    test('updateAppointment replaces the matching entry', () {
      state.addAppointment(Appointment(id: 'a1', title: 'Old title', dateTime: '', location: '', notes: ''));
      state.updateAppointment('a1', Appointment(id: 'a1', title: 'New title', dateTime: '', location: '', notes: ''));
      expect(state.appointments.first.title, 'New title');
    });

    test('deleteAppointment removes the matching entry', () {
      state.addAppointment(Appointment(id: 'a1', title: 'A', dateTime: '', location: '', notes: ''));
      state.deleteAppointment('a1');
      expect(state.appointments, isEmpty);
    });
  });

  group('AppState — check-in and activity', () {
    late AppState state;
    setUp(() => state = AppState());

    test('checkIn logs an entry exactly once even if called twice', () {
      state.checkIn();
      state.checkIn();
      expect(state.checkedIn, isTrue);
      expect(state.activity.where((a) => a.type == ActivityType.checkedIn), hasLength(1));
    });
  });

  group('AppState — messages', () {
    late AppState state;
    setUp(() => state = AppState());

    test('sendMessage inserts a new, already-read message at the front', () {
      state.sendMessage(from: 'Maria Thompson', to: 'Dr. Sharma', subject: 'Hi', body: 'Body text');
      expect(state.messages, hasLength(1));
      final msg = state.messages.first;
      expect(msg.from, 'Maria Thompson');
      expect(msg.read, isTrue);
      expect(msg.archived, isFalse);
    });

    test('unreadMessageCount counts only unread, non-archived messages', () {
      state.messages.addAll([
        Message(id: '1', from: 'A', to: 'Me', subject: 's', body: 'b', timestamp: 't', read: false),
        Message(id: '2', from: 'B', to: 'Me', subject: 's', body: 'b', timestamp: 't', read: true),
        Message(id: '3', from: 'C', to: 'Me', subject: 's', body: 'b', timestamp: 't', read: false, archived: true),
      ]);
      expect(state.unreadMessageCount, 1);
    });

    test('markMessageRead only ever moves unread -> read, never read -> unread', () {
      state.messages.add(Message(id: '1', from: 'A', to: 'Me', subject: 's', body: 'b', timestamp: 't', read: false));
      state.markMessageRead('1');
      expect(state.messages.first.read, isTrue);

      // Calling it again on an already-read message must not flip it back.
      state.markMessageRead('1');
      expect(state.messages.first.read, isTrue);
    });

    test('toggleMessageRead flips read state in both directions', () {
      state.messages.add(Message(id: '1', from: 'A', to: 'Me', subject: 's', body: 'b', timestamp: 't', read: false));
      state.toggleMessageRead('1');
      expect(state.messages.first.read, isTrue);
      state.toggleMessageRead('1');
      expect(state.messages.first.read, isFalse);
    });

    test('archiveMessage removes the message from the unread count without deleting it', () {
      state.messages.add(Message(id: '1', from: 'A', to: 'Me', subject: 's', body: 'b', timestamp: 't', read: false));
      state.archiveMessage('1');
      expect(state.messages, hasLength(1));
      expect(state.messages.first.archived, isTrue);
      expect(state.unreadMessageCount, 0);
    });

    test('deleteMessage removes the message entirely', () {
      state.messages.add(Message(id: '1', from: 'A', to: 'Me', subject: 's', body: 'b', timestamp: 't'));
      state.deleteMessage('1');
      expect(state.messages, isEmpty);
    });
  });

  group('AppState — settings and auth', () {
    late AppState state;
    setUp(() => state = AppState());

    test('setHandMode / setTheme / setFontScale update fields and notify listeners', () {
      var notifications = 0;
      state.addListener(() => notifications++);

      state.setHandMode(HandMode.left);
      state.setTheme(ThemeModeSetting.dark);
      state.setFontScale(FontScale.large);

      expect(state.handMode, HandMode.left);
      expect(state.theme, ThemeModeSetting.dark);
      expect(state.fontSize, FontScale.large);
      expect(notifications, 3);
    });

    test('signIn sets the user name; signOut resets session flags but keeps app data', () {
      state.signIn('Maria Thompson');
      expect(state.userName, 'Maria Thompson');

      state.setHandMode(HandMode.left);
      state.checkIn();
      state.addMedication(Medication(id: 'm1', name: 'A', dose: '', time: '', notes: ''));

      state.signOut();
      expect(state.userName, '');
      expect(state.handMode, HandMode.off);
      expect(state.checkedIn, isFalse);
      // Signing out is not a data wipe — Margaret's records persist.
      expect(state.medications, hasLength(1));
    });

    test('signIn relabels demo data addressed to the placeholder "Maria Thompson" to whoever signs in', () {
      // Seed data (as built in main.dart) is created before anyone has
      // signed in, so it hardcodes a placeholder name for appointments
      // assigned to "the app's user" and messages addressed to them. Once a
      // real name is known, those records should follow the actual
      // signed-in person instead of forever reading "Maria Thompson".
      state.appointments = [
        Appointment(id: 'a1', title: 'Checkup', dateTime: 't', location: 'l', notes: '', assignee: 'Maria Thompson'),
        // Not everything is assigned to the app's user -- an assignee that
        // was deliberately set to someone else must not be touched.
        Appointment(id: 'a2', title: 'Eye test', dateTime: 't', location: 'l', notes: '', assignee: 'Emma Thompson'),
      ];
      state.messages = [
        Message(id: 'm1', from: 'Dr. Sharma', to: 'Maria Thompson', subject: 's', body: 'b', timestamp: 't'),
        // A message the user already sent to someone else keeps its real
        // recipient.
        Message(id: 'm2', from: 'Maria Thompson', to: 'Dr. Sharma', subject: 's', body: 'b', timestamp: 't'),
      ];

      state.signIn('Dom Puller');

      expect(state.appointments.firstWhere((a) => a.id == 'a1').assignee, 'Dom Puller');
      expect(state.appointments.firstWhere((a) => a.id == 'a2').assignee, 'Emma Thompson');
      expect(state.messages.firstWhere((m) => m.id == 'm1').to, 'Dom Puller');
      expect(state.messages.firstWhere((m) => m.id == 'm2').to, 'Dr. Sharma');
    });

    test('toggleWidget flips a dashboard widget\'s enabled flag', () {
      state.dashboardWidgets = [DashboardWidget(id: 'w1', label: 'Widget', enabled: true)];
      state.toggleWidget('w1');
      expect(state.dashboardWidgets.first.enabled, isFalse);
    });
  });

  group('AppState — clock-dependent timestamps (injected clock)', () {
    // AppState takes an optional `now` clock so tests never depend on the
    // wall clock (a test asserting real-time output would be flaky right
    // around midnight/noon and impossible to write deterministically at
    // all). Each case below pins a specific DateTime and checks the exact
    // "h:mm am/pm" string that activity/message logging produces from it.
    test('formats midnight as 12:00 am', () {
      final state = AppState(now: () => DateTime(2026, 1, 1, 0, 0));
      state.checkIn();
      expect(state.activity.first.timestamp, '12:00 am');
    });

    test('formats noon as 12:00 pm', () {
      final state = AppState(now: () => DateTime(2026, 1, 1, 12, 0));
      state.checkIn();
      expect(state.activity.first.timestamp, '12:00 pm');
    });

    test('formats an afternoon hour with minute padding', () {
      final state = AppState(now: () => DateTime(2026, 1, 1, 13, 5));
      state.checkIn();
      expect(state.activity.first.timestamp, '1:05 pm');
    });

    test('formats a morning hour just before noon', () {
      final state = AppState(now: () => DateTime(2026, 1, 1, 11, 59));
      state.checkIn();
      expect(state.activity.first.timestamp, '11:59 am');
    });

    test('sendMessage stamps the message with the injected clock too', () {
      final state = AppState(now: () => DateTime(2026, 6, 1, 9, 15));
      state.sendMessage(from: 'Maria Thompson', to: 'Dr. Sharma', subject: 'Hi', body: 'Body');
      expect(state.messages.first.timestamp, '9:15 am');
    });
  });
}

File: mobile-rn/__tests__/state/AppState.test.ts
// Jest unit tests for the AppState reducer — the RN counterpart of the
// Flutter suite's app_state_test.dart. Drives the pure reducer directly, no
// React tree needed.

import {
  appReducer,
  createInitialState,
  formatClockTime,
  unreadMessageCount,
} from '../../src/state/AppState';
import type { AppStateData } from '../../src/state/AppState';
import { seededInitialState } from '../../src/state/seed';
import {
  ALERT_APPOINTMENT_TODAY,
  ALERT_MEDS_UNTAKEN,
  ALERT_NO_CHECK_IN,
} from '../../src/models/types';
import type { Medication, Appointment } from '../../src/models/types';

const fixedNow = () => new Date(2026, 8, 15, 9, 30); // Sep 15 2026, 9:30 am

function fresh(): AppStateData {
  return seededInitialState(fixedNow);
}

describe('formatClockTime', () => {
  it('formats midnight as 12:00 am', () => {
    expect(formatClockTime(new Date(2026, 0, 1, 0, 0))).toBe('12:00 am');
  });
  it('formats noon as 12:00 pm', () => {
    expect(formatClockTime(new Date(2026, 0, 1, 12, 0))).toBe('12:00 pm');
  });
  it('pads single-digit minutes', () => {
    expect(formatClockTime(new Date(2026, 0, 1, 9, 5))).toBe('9:05 am');
  });
  it('formats afternoon hours', () => {
    expect(formatClockTime(new Date(2026, 0, 1, 14, 45))).toBe('2:45 pm');
  });
});

describe('signIn', () => {
  it('records the user name', () => {
    const s = appReducer(fresh(), { type: 'signIn', name: 'Dorothy' });
    expect(s.userName).toBe('Dorothy');
  });

  it('relabels appointments assigned to the demo placeholder', () => {
    const s = appReducer(fresh(), { type: 'signIn', name: 'Dorothy' });
    const assigned = s.appointments.filter((a) => a.assignee === 'Dorothy');
    expect(assigned).toHaveLength(2); // a1 and a2 were "Maria Thompson"
  });

  it('leaves unassigned appointments alone', () => {
    const s = appReducer(fresh(), { type: 'signIn', name: 'Dorothy' });
    const eye = s.appointments.find((a) => a.id === 'a3');
    expect(eye?.assignee).toBe('');
  });

  it('relabels messages addressed to the demo placeholder', () => {
    const s = appReducer(fresh(), { type: 'signIn', name: 'Dorothy' });
    expect(s.messages.every((m) => m.to === 'Dorothy')).toBe(true);
  });
});

describe('signOut', () => {
  it('resets user-scoped state but keeps data', () => {
    let s = appReducer(fresh(), { type: 'signIn', name: 'Dorothy' });
    s = appReducer(s, { type: 'setHandMode', mode: 'left' });
    s = appReducer(s, { type: 'checkIn' });
    s = appReducer(s, { type: 'dismissAlert', id: ALERT_MEDS_UNTAKEN });

    s = appReducer(s, { type: 'signOut' });

    expect(s.userName).toBe('');
    expect(s.handMode).toBe('off');
    expect(s.checkedIn).toBe(false);
    expect(s.dismissedAlertIds).toEqual([]);
    // Data survives sign-out (matches the Flutter build).
    expect(s.medications.length).toBe(3);
  });
});

describe('checkIn', () => {
  it('sets checkedIn and logs activity', () => {
    const s = appReducer(fresh(), { type: 'checkIn' });
    expect(s.checkedIn).toBe(true);
    expect(s.activity[0].type).toBe('checkedIn');
    expect(s.activity[0].description).toBe('Margaret checked in');
    expect(s.activity[0].timestamp).toBe('9:30 am'); // from the injected clock
  });

  it('is idempotent — a second check-in changes nothing', () => {
    let s = appReducer(fresh(), { type: 'checkIn' });
    const afterFirst = s.activity.length;
    s = appReducer(s, { type: 'checkIn' });
    expect(s.activity.length).toBe(afterFirst);
  });

  it('clears a previously dismissed no-check-in alert', () => {
    let s = appReducer(fresh(), { type: 'dismissAlert', id: ALERT_NO_CHECK_IN });
    expect(s.dismissedAlertIds).toContain(ALERT_NO_CHECK_IN);
    s = appReducer(s, { type: 'checkIn' });
    expect(s.dismissedAlertIds).not.toContain(ALERT_NO_CHECK_IN);
  });
});

describe('medications', () => {
  const newMed: Medication = {
    id: 'm9',
    name: 'Aspirin',
    dose: '81 mg — 1 tablet',
    time: '8:00 am',
    notes: '',
    taken: false,
  };

  it('addMedication appends and logs activity', () => {
    const s = appReducer(fresh(), { type: 'addMedication', med: newMed });
    expect(s.medications).toHaveLength(4);
    expect(s.medications[3].name).toBe('Aspirin');
    expect(s.activity[0].description).toBe("Added Aspirin to Margaret's medications");
  });

  it('addMedication re-surfaces a dismissed meds-untaken alert', () => {
    let s = appReducer(fresh(), { type: 'dismissAlert', id: ALERT_MEDS_UNTAKEN });
    s = appReducer(s, { type: 'addMedication', med: newMed });
    expect(s.dismissedAlertIds).not.toContain(ALERT_MEDS_UNTAKEN);
  });

  it('updateMedication replaces in place by id', () => {
    const updated = { ...newMed, id: 'm1', name: 'Amlodipine XL' };
    const s = appReducer(fresh(), { type: 'updateMedication', id: 'm1', updated });
    expect(s.medications.find((m) => m.id === 'm1')?.name).toBe('Amlodipine XL');
    expect(s.medications).toHaveLength(3);
  });

  it('deleteMedication removes by id', () => {
    const s = appReducer(fresh(), { type: 'deleteMedication', id: 'm2' });
    expect(s.medications.map((m) => m.id)).toEqual(['m1', 'm3']);
  });

  it('toggleMedTaken flips taken and logs medicationTaken', () => {
    const s = appReducer(fresh(), { type: 'toggleMedTaken', id: 'm1' });
    expect(s.medications.find((m) => m.id === 'm1')?.taken).toBe(true);
    expect(s.activity[0].type).toBe('medicationTaken');
    expect(s.activity[0].description).toBe('Amlodipine marked as taken');
  });

  it('toggleMedTaken back off logs medicationUnmarked', () => {
    const s = appReducer(fresh(), { type: 'toggleMedTaken', id: 'm2' }); // m2 starts taken
    expect(s.medications.find((m) => m.id === 'm2')?.taken).toBe(false);
    expect(s.activity[0].type).toBe('medicationUnmarked');
    expect(s.activity[0].description).toBe('Metformin unmarked');
  });

  it('toggleMedTaken on an unknown id is a no-op', () => {
    const before = fresh();
    const s = appReducer(before, { type: 'toggleMedTaken', id: 'nope' });
    expect(s).toBe(before);
  });
});

describe('appointments', () => {
  const newAppt: Appointment = {
    id: 'a9',
    title: 'Physio session',
    dateTime: 'Tomorrow — 9:00 am',
    location: 'Westfield Clinic',
    assignee: '',
    notes: '',
  };

  it('addAppointment inserts at the front', () => {
    const s = appReducer(fresh(), { type: 'addAppointment', appt: newAppt });
    expect(s.appointments[0].title).toBe('Physio session');
    expect(s.appointments).toHaveLength(4);
  });

  it('addAppointment re-surfaces a dismissed appointment-today alert', () => {
    let s = appReducer(fresh(), { type: 'dismissAlert', id: ALERT_APPOINTMENT_TODAY });
    s = appReducer(s, { type: 'addAppointment', appt: newAppt });
    expect(s.dismissedAlertIds).not.toContain(ALERT_APPOINTMENT_TODAY);
  });

  it('updateAppointment replaces by id', () => {
    const updated = { ...newAppt, id: 'a2', title: 'Rescheduled review' };
    const s = appReducer(fresh(), { type: 'updateAppointment', id: 'a2', updated });
    expect(s.appointments.find((a) => a.id === 'a2')?.title).toBe('Rescheduled review');
  });

  it('deleteAppointment removes by id', () => {
    const s = appReducer(fresh(), { type: 'deleteAppointment', id: 'a1' });
    expect(s.appointments.map((a) => a.id)).toEqual(['a2', 'a3']);
  });
});

describe('messages', () => {
  it('sendMessage prepends a read message with a clock timestamp', () => {
    const s = appReducer(
      fresh(),
      { type: 'sendMessage', from: 'Maria', to: 'Dr. Sharma', subject: 'Thanks', body: 'Got it.' },
    );
    expect(s.messages[0].subject).toBe('Thanks');
    expect(s.messages[0].read).toBe(true);
    expect(s.messages[0].timestamp).toBe('9:30 am');
  });

  it('markMessageRead flips unread to read only', () => {
    let s = fresh();
    expect(s.messages.find((m) => m.id === 'msg1')?.read).toBe(false);
    s = appReducer(s, { type: 'markMessageRead', id: 'msg1' });
    expect(s.messages.find((m) => m.id === 'msg1')?.read).toBe(true);
    // Already-read stays read.
    s = appReducer(s, { type: 'markMessageRead', id: 'msg1' });
    expect(s.messages.find((m) => m.id === 'msg1')?.read).toBe(true);
  });

  it('toggleMessageRead flips both directions', () => {
    let s = appReducer(fresh(), { type: 'toggleMessageRead', id: 'msg3' }); // starts read
    expect(s.messages.find((m) => m.id === 'msg3')?.read).toBe(false);
    s = appReducer(s, { type: 'toggleMessageRead', id: 'msg3' });
    expect(s.messages.find((m) => m.id === 'msg3')?.read).toBe(true);
  });

  it('archiveMessage hides from the inbox; unarchiveMessage restores', () => {
    let s = appReducer(fresh(), { type: 'archiveMessage', id: 'msg1' });
    expect(s.messages.find((m) => m.id === 'msg1')?.archived).toBe(true);
    s = appReducer(s, { type: 'unarchiveMessage', id: 'msg1' });
    expect(s.messages.find((m) => m.id === 'msg1')?.archived).toBe(false);
  });

  it('deleteMessage removes by id', () => {
    const s = appReducer(fresh(), { type: 'deleteMessage', id: 'msg2' });
    expect(s.messages.map((m) => m.id)).toEqual(['msg1', 'msg3']);
  });

  it('unreadMessageCount counts unread non-archived only', () => {
    let s = fresh(); // msg1 + msg2 unread, msg3 read
    expect(unreadMessageCount(s.messages)).toBe(2);
    s = appReducer(s, { type: 'archiveMessage', id: 'msg1' });
    expect(unreadMessageCount(s.messages)).toBe(1);
    s = appReducer(s, { type: 'markMessageRead', id: 'msg2' });
    expect(unreadMessageCount(s.messages)).toBe(0);
  });
});

describe('dashboard widgets', () => {
  it('toggleWidget flips enabled for the matching id only', () => {
    const s = appReducer(fresh(), { type: 'toggleWidget', id: 'alerts' });
    expect(s.dashboardWidgets.find((w) => w.id === 'alerts')?.enabled).toBe(false);
    expect(s.dashboardWidgets.find((w) => w.id === 'status')?.enabled).toBe(true);
  });

  it('reorderWidgets rewrites order fields to match the new sequence', () => {
    const s = fresh();
    const reordered = [...s.dashboardWidgets].reverse();
    const next = appReducer(s, { type: 'reorderWidgets', ordered: reordered });
    expect(next.dashboardWidgets.map((w) => w.order)).toEqual([0, 1, 2, 3, 4]);
    expect(next.dashboardWidgets[0].id).toBe('messages');
  });
});

describe('settings', () => {
  it('setHandMode / setTheme / setFontScale update their fields', () => {
    let s = appReducer(fresh(), { type: 'setHandMode', mode: 'left' });
    expect(s.handMode).toBe('left');
    s = appReducer(s, { type: 'setTheme', theme: 'dark' });
    expect(s.theme).toBe('dark');
    s = appReducer(s, { type: 'setFontScale', scale: 'xlarge' });
    expect(s.fontSize).toBe('xlarge');
  });
});

describe('createInitialState', () => {
  it('starts signed out with empty collections', () => {
    const s = createInitialState(fixedNow);
    expect(s.userName).toBe('');
    expect(s.medications).toEqual([]);
    expect(s.messages).toEqual([]);
    expect(s.theme).toBe('system');
    expect(s.handMode).toBe('off');
  });
});
